### PR TITLE
Adds Bonfire Grills

### PR DESCRIPTION
**Pull request by fludd12**
Using metal rods on a bonfire will now give you the option of adding a stake or a grill. Adding a grill lets you cook things on the bonfire, but not with the same level of control as a microwave.

Also this is a complete codersprite I apologize.
codersprite

:cl: fludd12
add: You can now add grills to a bonfire, letting you cook things on top of them.
/:cl:

### CODE REVIEW

**Comment by FantasticFwoosh** [comment]
Sounds good.

**Comment by Grazyn** [comment]
We don't have marshmallows though. I don't care if it's codersprites, someone please add them

**Comment by FantasticFwoosh** [comment]
Does this heat up to a 1000c? because that's the amount of heat you need to cook a goliath steak and also the max (the ability to put reagent holding beakers onto on it too would be nice) heat that the microwave cooks at.

Might make a arguement to remove the bonfires O2 restrictions as well as remove throwing goliath steaks onto lava to cook in favour of having a BBQ.

**Comment by Kromgar** [comment]
Wait.... how do you light on fire if there is no oxygen on lavaland?

**Comment by Jazaen** [comment]
Space magic. People can be caught on fire on lavaland.

**Comment by FantasticFwoosh** [comment]
See in reference  `/obj/structure/bonfire/proc/CheckOxygen()` at towercap.dm where the bonfire code is kept, there were references to make everything else that burns on lavaland respect those rules too (torches, cigarettes etc), ill find the PR if i can.

**Comment by fludd12** [comment]
To answer the questions, I was thinking about making it able to cook goliath steaks, but decided that it would make sense for something that slow-cooks to not be able to accomplish the same effect as a quick dip in lava.
As for how it cooks, it calls each `obj/item`'s `microwave_act` on a random timer, so while it can cook things, it can also _over_cook things if you aren't careful.
Mostly, I wanted this so that podpeople can prepare banquets for when they finally meet their masters.

File: code/modules/hydroponics/grown/towercap.dm [hunk 1]
```diff
@@ -107,20 +107,44 @@
...
 		var/obj/item/stack/rods/R = W
-		R.use(1)
-		can_buckle = 1
-		buckle_requires_restraints = 1
-		to_chat(user, "<span class='italics'>You add a rod to [src].")
-		var/mutable_appearance/rod_underlay = mutable_appearance('icons/obj/hydroponics/equipment.dmi', "bonfire_rod")
-		rod_underlay.pixel_y = 16
-		underlays += rod_underlay
+		var/choice = input(user, "What would you like to construct?", "Bonfire") as null|anything in list("Stake","Grill")
+		if(choice == "Stake")
```
Comment (lzimann) [inline code/modules/hydroponics/grown/towercap.dm]: Use a switch

File: code/modules/hydroponics/grown/towercap.dm [hunk 2]
```diff
@@ -107,20 +107,44 @@
...
+			R.use(1)
+			can_buckle = 1
```
Comment (lzimann) [inline code/modules/hydroponics/grown/towercap.dm]: `TRUE`

File: code/modules/hydroponics/grown/towercap.dm [hunk 3]
```diff
@@ -107,20 +107,44 @@
...
+			buckle_requires_restraints = 1
```
Comment (lzimann) [inline code/modules/hydroponics/grown/towercap.dm]: `TRUE`

File: code/modules/hydroponics/grown/towercap.dm [hunk 4]
```diff
@@ -107,20 +107,44 @@
...
+			to_chat(user, "<span class='italics'>You add a rod to [src].")
+			var/mutable_appearance/rod_underlay = mutable_appearance('icons/obj/hydroponics/equipment.dmi', "bonfire_rod")
+			rod_underlay.pixel_y = 16
+			underlays += rod_underlay
+		if(choice == "Grill")
+			R.use(1)
+			grill = TRUE
+			to_chat(user, "<span class='italics'>You add a grill to [src].")
+			grill = 1
```
Comment (lzimann) [inline code/modules/hydroponics/grown/towercap.dm]: Repeating grill here

File: code/modules/hydroponics/grown/towercap.dm [hunk 5]
```diff
@@ -107,20 +107,44 @@
...
+			if(user.drop_item())
+				W.Move(get_turf(src))
+				var/list/click_params = params2list(params)
+				//Center the icon where the user clicked.
+				if(!click_params || !click_params["icon-x"] || !click_params["icon-y"])
+					return
+				//Clamp it so that the icon never moves more than 16 pixels in either direction (thus leaving the table turf)
+				W.pixel_x = Clamp(text2num(click_params["icon-x"]) - 16, -(world.icon_size/2), world.icon_size/2)
+				W.pixel_y = Clamp(text2num(click_params["icon-y"]) - 16, -(world.icon_size/2), world.icon_size/2)
+				return 1
```
Comment (lzimann) [inline code/modules/hydroponics/grown/towercap.dm]: `TRUE`

File: code/modules/hydroponics/grown/towercap.dm [hunk 6]
```diff
@@ -175,11 +199,27 @@
...
+/obj/structure/bonfire/proc/Cook()
+	var/turf/current_location = get_turf(src)
+	for(var/A in current_location)
+		if(A == src)
+			continue
+		else if(isliving(A)) //It's still a fire, idiot.
+			var/mob/living/L = A
+			L.adjust_fire_stacks(fire_stack_strength)
+			L.IgniteMob()
+	for(var/obj/item/O in current_location)
```
Comment (lzimann) [inline code/modules/hydroponics/grown/towercap.dm]: Just include this 2nd loop in the first with a typecheck + typecasting. 
```
else if(istype(A, /obj/item) && prob(20))
    var/obj/item/O = A
    O.microwave_act()
```

**Review by lzimann — CHANGES_REQUESTED** [review]

File: code/modules/hydroponics/grown/towercap.dm [hunk 7]
```diff
@@ -107,20 +107,44 @@
...
+			grill = TRUE
+			to_chat(user, "<span class='italics'>You add a grill to [src].")
+			grill = 1
+			var/mutable_appearance/grill_overlay = mutable_appearance('icons/obj/hydroponics/equipment.dmi', "bonfire_grill")
+			overlays += grill_overlay
 	if(W.is_hot())
 		StartBurning()
+	if(grill)
+		if(user.a_intent != INTENT_HARM && !(W.flags & ABSTRACT))
+			if(user.drop_item())
```
Comment (lzimann) [inline code/modules/hydroponics/grown/towercap.dm]: @Cyberboss What is the proper inventory proc for this again?

File: code/modules/hydroponics/grown/towercap.dm [hunk 8]
```diff
@@ -107,20 +107,44 @@
...
-		to_chat(user, "<span class='italics'>You add a rod to [src].")
-		var/mutable_appearance/rod_underlay = mutable_appearance('icons/obj/hydroponics/equipment.dmi', "bonfire_rod")
-		rod_underlay.pixel_y = 16
-		underlays += rod_underlay
+		var/choice = input(user, "What would you like to construct?", "Bonfire") as null|anything in list("Stake","Grill")
+		if(choice == "Stake")
+			R.use(1)
+			can_buckle = 1
+			buckle_requires_restraints = 1
+			to_chat(user, "<span class='italics'>You add a rod to [src].")
```
Comment (Cyberboss) [inline code/modules/hydroponics/grown/towercap.dm]: \the [src]

File: code/modules/hydroponics/grown/towercap.dm (hunk 7 above)
Comment (Cyberboss) [inline code/modules/hydroponics/grown/towercap.dm]: use user.temporarilyRemoveItemFromInventory(W)

File: code/modules/hydroponics/grown/towercap.dm [hunk 9]
```diff
@@ -107,20 +107,44 @@
...
+			to_chat(user, "<span class='italics'>You add a grill to [src].")
+			grill = 1
+			var/mutable_appearance/grill_overlay = mutable_appearance('icons/obj/hydroponics/equipment.dmi', "bonfire_grill")
+			overlays += grill_overlay
 	if(W.is_hot())
 		StartBurning()
+	if(grill)
+		if(user.a_intent != INTENT_HARM && !(W.flags & ABSTRACT))
+			if(user.drop_item())
+				W.Move(get_turf(src))
```
Comment (Cyberboss) [inline code/modules/hydroponics/grown/towercap.dm]: W.forceMove(get_turf(src))

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by fludd12** [comment]
Heh, is it obvious that I salvaged a lot of this code from other sources? Thanks for all the help.

**Comment by fludd12** [comment]
Alright, that should do it, unless anyone else has any items of note?

**Review by lzimann — APPROVED** [review]

**Pull request merged**